File: Untitled.kt
/**
 * VulnLabKtor.kt
 *
 * Single-file Ktor app (vulnerable endpoints) — LOCAL-ONLY & TOKEN-GATED
 *
 * WARNING: Run only in an isolated VM/container bound to localhost.
 *
 * How to run (quick):
 * 1) Create a Gradle Kotlin project (minimal). In build.gradle.kts include:
 *
 *    plugins {
 *      kotlin("jvm") version "1.9.0"
 *      application
 *    }
 *    repositories { mavenCentral() }
 *    dependencies {
 *      implementation("io.ktor:ktor-server-netty:2.3.5")
 *      implementation("io.ktor:ktor-server-core:2.3.5")
 *      implementation("io.ktor:ktor-server-cio:2.3.5")
 *      implementation("io.ktor:ktor-server-host-common:2.3.5")
 *      implementation("io.ktor:ktor-server-html-builder:2.3.5")
 *      implementation("org.xerial:sqlite-jdbc:3.42.0.0")
 *      implementation("ch.qos.logback:logback-classic:1.4.7")
 *    }
 *    application { mainClass.set("VulnLabKtorKt") }
 *
 * 2) Put this file as src/main/kotlin/VulnLabKtor.kt
 * 3) Run `./gradlew run` inside an isolated VM/container bound to localhost.
 * 4) Open http://127.0.0.1:8080/ in the same VM. Root will show the token.
 *
 * NOTE: This app intentionally contains vulnerabilities for testing scanners. Do NOT expose to public networks.
 */

import io.ktor.server.engine.*
import io.ktor.server.netty.*
import io.ktor.server.application.*
import io.ktor.server.request.*
import io.ktor.server.response.*
import io.ktor.server.routing.*
import io.ktor.server.http.content.*
import io.ktor.server.sessions.*
import io.ktor.util.pipeline.*
import io.ktor.http.*
import io.ktor.http.content.*
import java.sql.DriverManager
import java.sql.Connection
import java.io.File
import kotlin.system.exitProcess
import kotlin.text.Charsets
import java.security.SecureRandom
import java.util.Base64
import java.nio.file.Files
import java.nio.file.Paths
import java.lang.ProcessBuilder

fun main() {
    // Auto-generate token if not set
    val envToken = System.getenv("VULN_TOKEN")?.takeIf { it.isNotBlank() } ?: run {
        val rnd = SecureRandom()
        val b = ByteArray(16)
        rnd.nextBytes(b)
        val gen = b.joinToString("") { "%02x".format(it) }
        // Not permanently setting environment variables in JVM; but we'll use this value in appConfig below
        gen
    }

    // Database file in the working directory
    val dbPath = "vulnlab_ktor.db"
    Class.forName("org.sqlite.JDBC")
    val conn = DriverManager.getConnection("jdbc:sqlite:$dbPath")
    conn.createStatement().use { st ->
        st.executeUpdate("PRAGMA journal_mode=WAL;")
        st.executeUpdate("CREATE TABLE IF NOT EXISTS Users(Id INTEGER PRIMARY KEY, Username TEXT, Password TEXT);")
        st.executeUpdate("CREATE TABLE IF NOT EXISTS Notes(Id INTEGER PRIMARY KEY, Owner TEXT, Content TEXT);")
        st.executeUpdate("CREATE TABLE IF NOT EXISTS Files(Id INTEGER PRIMARY KEY, Owner TEXT, Filename TEXT, Path TEXT);")
        // ensure an admin user exists (plaintext password)
        st.executeUpdate("INSERT OR IGNORE INTO Users(Username, Password) VALUES('admin','123');")
    }

    // ensure uploads dir exists
    val uploadDir = Paths.get("uploads")
    if (!Files.exists(uploadDir)) Files.createDirectories(uploadDir)

    // Launch Ktor embedded server (Netty)
    embeddedServer(Netty, port = 8080, host = "127.0.0.1") {
        module(envToken, dbPath, uploadDir.toString())
    }.start(wait = true)
}

fun Application.module(token: String, dbPath: String, uploadDir: String) {
    // Middleware: allow root only without token so user can see token; otherwise require local-only + token
    intercept(ApplicationCallPipeline.Plugins) {
        val path = call.request.path()
        if (path == "/") return@intercept
        if (!isLocal(call)) {
            call.respond(HttpStatusCode.Forbidden, "Access denied: this lab accepts requests from localhost only.")
            finish() // stop processing
            return@intercept
        }
        val reqToken = call.request.queryParameters["token"] ?: ""
        if (reqToken != token) {
            call.respond(HttpStatusCode.Forbidden, "Access denied: missing or invalid token. See root / for token.")
            finish()
            return@intercept
        }
    }

    routing {
        // Root displays token and tokenized links
        get("/") {
            val html = buildString {
                append("<h2>VulnLab Ktor (LOCAL ONLY)</h2>")
                append("<p><strong>Auto token:</strong> <code>${token}</code></p>")
                append("<p>Click: <a href=\"/home?token=${token}\">Open lab home (local)</a></p><hr/>")
                append("<ul>")
                append("<li><a href=\"/home?token=${token}\">Home</a></li>")
                append("<li><a href=\"/notes?token=${token}\">A01 Broken Access Control</a></li>")
                append("<li><a href=\"/crypto?token=${token}\">A02 Cryptographic Failures</a></li>")
                append("<li><a href=\"/login?user=admin&pass=123&token=${token}\">A03 Injection (login)</a></li>")
                append("<li><a href=\"/transfer?from=a&to=b&amount=5000&token=${token}\">A04 Insecure Design</a></li>")
                append("<li><a href=\"/config?token=${token}\">A05 Misconfiguration</a></li>")
                append("<li><a href=\"/components?token=${token}\">A06 Outdated Components</a></li>")
                append("<li><form method='post' action='/register?token=${token}'><input name='user' placeholder='user'/> <input name='pass' placeholder='pass'/> <button>Register (A07)</button></form></li>")
                append("<li><a href=\"/update?token=${token}\">A08 Integrity Failures</a></li>")
                append("<li><a href=\"/log?token=${token}\">A09 Logging Failures</a></li>")
                append("<li><a href=\"/fetch?url=http://127.0.0.1:8080/config&token=${token}\">A10 SSRF</a></li>")
                append("<li><a href=\"/upload?token=${token}\">Upload (Unrestricted)</a></li>")
                append("</ul>")
            }
            call.respondText(html, ContentType.Text.Html)
        }

        get("/home") {
            val t = call.request.queryParameters["token"] ?: token
            val html = """
                <h3>Vuln Lab Home</h3>
                <ul>
                  <li><a href="/notes?token=$t">Notes</a></li>
                  <li><a href="/crypto?token=$t">Crypto</a></li>
                  <li><a href="/login?user=admin&pass=123&token=$t">Login</a></li>
                  <li><a href="/upload?token=$t">Upload</a></li>
                </ul>
            """.trimIndent()
            call.respondText(html, ContentType.Text.Html)
        }

        // -------------------------------
        // A01 Broken Access Control + Stored XSS
        // -------------------------------
        get("/notes") {
            val sb = StringBuilder("<h3>All Notes (no ACL)</h3>")
            DriverManager.getConnection("jdbc:sqlite:$dbPath").use { c ->
                val st = c.createStatement()
                val rs = st.executeQuery("SELECT Id, Owner, Content FROM Notes;")
                while (rs.next()) {
                    val owner = rs.getString("Owner")
                    val content = rs.getString("Content")
                    // intentionally NOT escaping -> stored XSS possible
                    sb.append("<div style='border:1px solid #ccc;margin:6px;padding:6px'><strong>Owner:</strong> $owner<br/><strong>Content:</strong> $content</div>")
                }
            }
            call.respondText(sb.toString(), ContentType.Text.Html)
        }

        post("/addnote") {
            val params = call.receiveParameters()
            val owner = params["owner"] ?: "anonymous"
            val content = params["content"] ?: ""
            // insecure: concatenated SQL -> SQLi possible
            DriverManager.getConnection("jdbc:sqlite:$dbPath").use { c ->
                val st = c.createStatement()
                st.executeUpdate("INSERT INTO Notes(Owner, Content) VALUES('$owner', '$content');")
            }
            call.respondText("Note added (stored XSS possible)", ContentType.Text.Plain)
        }

        // -------------------------------
        // A02 Crypto (plaintext + weak IV)
        // -------------------------------
        get("/crypto") {
            val secret = "SensitivePlaintext"
            // weak key + predictable IV demonstration (do NOT use)
            val key = "weakkeyweakkey" // 16 bytes
            val iv = ByteArray(16) // zeros
            val cipherHex = try {
                // naive XOR "encryption" as weak placeholder for demo
                val cipher = secret.toByteArray(Charsets.UTF_8).mapIndexed { i, b -> (b.toInt() xor key[i % key.length].code).toByte() }.toByteArray()
                cipher.joinToString("") { "%02x".format(it) }
            } catch (e: Exception) { "encrypt-failed" }
            call.respondText("Plaintext: $secret<br/>Weak cipher hex: $cipherHex", ContentType.Text.Html)
        }

        // -------------------------------
        // A03 Injection (SQLi) - login
        // -------------------------------
        get("/login") {
            val user = call.request.queryParameters["user"] ?: ""
            val pass = call.request.queryParameters["pass"] ?: ""
            var count = 0
            DriverManager.getConnection("jdbc:sqlite:$dbPath").use { c ->
                val st = c.createStatement()
                // insecure concatenation → SQLi
                val rs = st.executeQuery("SELECT COUNT(*) AS c FROM Users WHERE Username='$user' AND Password='$pass';")
                if (rs.next()) count = rs.getInt("c")
            }
            call.respondText(if (count == 1) "Login ok" else "Invalid", ContentType.Text.Plain)
        }

        // -------------------------------
        // A04 Insecure Design (no auth / limits)
        // -------------------------------
        get("/transfer") {
            val from = call.request.queryParameters["from"] ?: "a"
            val to = call.request.queryParameters["to"] ?: "b"
            val amount = call.request.queryParameters["amount"] ?: "0"
            call.respondText("Transferred $amount from $from to $to", ContentType.Text.Plain)
        }

        // -------------------------------
        // A05 Security Misconfiguration / Info leak
        // -------------------------------
        get("/config") {
            val env = System.getenv().map { (k, v) -> "$k=$v" }.joinToString("\n")
            val stack = Thread.currentThread().stackTrace.joinToString("\n")
            call.respondText("<pre>Environment:\n$env\n\nStack:\n$stack</pre>", ContentType.Text.Html)
        }

        // -------------------------------
        // A06 Outdated components (simulated)
        // -------------------------------
        get("/components") {
            call.respondText("Simulated: outdated components (demo)", ContentType.Text.Plain)
        }

        // -------------------------------
        // A07 Auth failures - register stores plaintext password
        // -------------------------------
        post("/register") {
            val params = call.receiveParameters()
            val user = params["user"] ?: ""
            val pass = params["pass"] ?: ""
            if (user.isBlank() || pass.isBlank()) {
                call.respond(HttpStatusCode.BadRequest, "user/pass required")
                return@post
            }
            DriverManager.getConnection("jdbc:sqlite:$dbPath").use { c ->
                val st = c.createStatement()
                // insecure concatenation → SQLi possible
                st.executeUpdate("INSERT INTO Users(Username, Password) VALUES('$user', '$pass');")
            }
            // insecure cookie (no HttpOnly/Secure)
            call.response.cookies.append("lab_user", user)
            call.respondText("User created (plaintext password stored)", ContentType.Text.Plain)
        }

        // -------------------------------
        // A08 Integrity failures - update (no signature verification)
        // -------------------------------
        get("/update") {
            call.respondText("Downloading updates from unverified source...", ContentType.Text.Plain)
        }

        // -------------------------------
        // A09 Logging & Monitoring failures (show exception)
        // -------------------------------
        get("/log") {
            try {
                throw RuntimeException("Example debug exception")
            } catch (e: Exception) {
                // insecure: show exception details to user
                call.respondText("Error: ${e}\n${e.stackTrace.joinToString("\n")}", ContentType.Text.Plain)
            }
        }

        // -------------------------------
        // A10 SSRF - fetch arbitrary URL
        // -------------------------------
        get("/fetch") {
            val url = call.request.queryParameters["url"] ?: ""
            if (url.isBlank()) {
                call.respond(HttpStatusCode.BadRequest, "Provide ?url=")
                return@get
            }
            val content = try {
                java.net.URL(url).readText()
            } catch (e: Exception) {
                "Fetch failed: ${e.message}"
            }
            // reflect raw content unsafely (possible XSS)
            call.respondText(content, ContentType.Text.Html)
        }

        // -------------------------------
        // Unrestricted file upload & download
        // -------------------------------
        get("/upload") {
            val t = call.request.queryParameters["token"] ?: token
            val sb = StringBuilder()
            sb.append("<h3>Upload (Unrestricted)</h3>")
            sb.append("<form method='post' enctype='multipart/form-data' action='/upload?token=$t'>")
            sb.append("Owner: <input name='owner'/><br/>File: <input type='file' name='file'/><br/><button>Upload</button></form>")
            sb.append("<h4>Files</h4><ul>")
            val files = File(uploadDir).listFiles() ?: arrayOf()
            for (f in files) {
                sb.append("<li>${f.name} - <a href='/download?name=${f.name}&token=$t'>download</a></li>")
            }
            sb.append("</ul>")
            call.respondText(sb.toString(), ContentType.Text.Html)
        }

        post("/upload") {
            val multipart = call.receiveMultipart()
            var owner = "anonymous"
            var savedName: String? = null
            multipart.forEachPart { part ->
                when (part) {
                    is PartData.FormItem -> {
                        if (part.name == "owner") owner = part.value
                    }
                    is PartData.FileItem -> {
                        val orig = part.originalFileName ?: "upload.bin"
                        val sanitized = Paths.get(orig).fileName.toString() // remove path
                        val dest = Paths.get(uploadDir, sanitized).toFile()
                        part.streamProvider().use { its ->
                            dest.outputStream().buffered().use { fos -> its.copyTo(fos) }
                        }
                        savedName = sanitized
                        // insecure DB insertion with concatenation
                        DriverManager.getConnection("jdbc:sqlite:$dbPath").use { c ->
                            val st = c.createStatement()
                            st.executeUpdate("INSERT INTO Files(Owner, Filename, Path) VALUES('$owner', '$sanitized', '${dest.absolutePath}');")
                        }
                    }
                    else -> {}
                }
                part.dispose()
            }
            call.respondText("Uploaded as ${savedName ?: "unknown"}", ContentType.Text.Plain)
        }

        get("/download") {
            val name = call.request.queryParameters["name"] ?: ""
            if (name.isBlank()) {
                call.respond(HttpStatusCode.BadRequest, "Provide ?name=")
                return@get
            }
            val f = File(Paths.get(uploadDir, name).toString())
            if (!f.exists()) {
                call.respond(HttpStatusCode.NotFound, "file not found")
                return@get
            }
            call.respondFile(f)
        }

        // -------------------------------
        // Command execution (dangerous)
        // -------------------------------
        get("/cmd") {
            val c = call.request.queryParameters["c"] ?: ""
            if (c.isBlank()) {
                call.respond(HttpStatusCode.BadRequest, "Provide ?c=")
                return@get
            }
            try {
                // insecure: direct shell execution
                val pb = ProcessBuilder("/bin/sh", "-c", c)
                val proc = pb.start()
                val out = proc.inputStream.bufferedReader().readText()
                val err = proc.errorStream.bufferedReader().readText()
                proc.waitFor()
                call.respondText("OUT:\n$out\nERR:\n$err", ContentType.Text.Plain)
            } catch (e: Exception) {
                call.respondText("Exec failed: ${e.message}", ContentType.Text.Plain)
            }
        }

        // Reflected XSS demo
        get("/echo") {
            val q = call.request.queryParameters["q"] ?: ""
            // unsafe reflect
            call.respondText("You said: $q", ContentType.Text.Html)
        }

        // Simple search demonstrating SQLi via LIKE concatenation
        get("/search") {
            val q = call.request.queryParameters["q"] ?: ""
            val sb = StringBuilder("<h3>Search results</h3>")
            DriverManager.getConnection("jdbc:sqlite:$dbPath").use { c ->
                val st = c.createStatement()
                val rs = st.executeQuery("SELECT Username FROM Users WHERE Username LIKE '%$q%';")
                while (rs.next()) {
                    sb.append(rs.getString("Username") + "<br/>")
                }
            }
            call.respondText(sb.toString(), ContentType.Text.Html)
        }

    } // routing end
}

// helper to detect local requests
fun isLocal(call: ApplicationCall): Boolean {
    val addr = call.request.host() // returns host header, not ideal; better to use remoteHost from request
    // Ktor provides remoteHost via request.local/remote? Use X-Forwarded-For risks — we check socket
    val remote = call.request.origin.remoteHost
    return remote == "127.0.0.1" || remote == "0:0:0:0:0:0:0:1" || remote == "localhost" || remote == "::1"
}

File: Untitled.swift
import Vapor
import Fluent
import FluentSQLiteDriver
import Foundation
import Crypto

// ---------- Configuration ----------
let app = Application(.development)
defer { app.shutdown() }

// Bind to localhost only for safety
app.http.server.configuration.hostname = "127.0.0.1"
app.http.server.configuration.port = 8080

// Auto-generate VULN_TOKEN if not set
let envToken = Environment.get("VULN_TOKEN")
let vulnToken: String = {
    if let t = envToken, !t.isEmpty { return t }
    // generate 32 hex chars token
    var bytes = [UInt8](repeating: 0, count: 16)
    _ = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
    let token = bytes.map { String(format: "%02x", $0) }.joined()
    setenv("VULN_TOKEN", token, 1)
    return token
}()

// ---------- Database (SQLite) ----------
let dbPath = app.directory.workingDirectory + "vulnlab.sqlite"
app.databases.use(.sqlite(.file(dbPath)), as: .sqlite)

// ---------- Models ----------
final class User: Model, Content {
    static let schema = "users"
    @ID(key: .id) var id: UUID?
    @Field(key: "username") var username: String
    @Field(key: "password") var password: String // plaintext intentionally
    init() {}
    init(id: UUID? = nil, username: String, password: String) {
        self.id = id; self.username = username; self.password = password
    }
}

final class Note: Model, Content {
    static let schema = "notes"
    @ID(key: .id) var id: UUID?
    @Field(key: "owner") var owner: String
    @Field(key: "content") var content: String
    init() {}
    init(owner: String, content: String) {
        self.owner = owner; self.content = content
    }
}

final class StoredFile: Model, Content {
    static let schema = "files"
    @ID(key: .id) var id: UUID?
    @Field(key: "owner") var owner: String
    @Field(key: "filename") var filename: String
    @Field(key: "path") var path: String
    init() {}
    init(owner: String, filename: String, path: String) {
        self.owner = owner; self.filename = filename; self.path = path
    }
}

// ---------- Migrations ----------
struct CreateUser: Migration {
    func prepare(on database: Database) -> EventLoopFuture<Void> {
        database.schema(User.schema)
            .id()
            .field("username", .string, .required)
            .field("password", .string, .required)
            .unique(on: "username")
            .create()
    }
    func revert(on database: Database) -> EventLoopFuture<Void> {
        database.schema(User.schema).delete()
    }
}
struct CreateNote: Migration {
    func prepare(on database: Database) -> EventLoopFuture<Void> {
        database.schema(Note.schema)
            .id()
            .field("owner", .string, .required)
            .field("content", .string, .required)
            .create()
    }
    func revert(on database: Database) -> EventLoopFuture<Void> {
        database.schema(Note.schema).delete()
    }
}
struct CreateFile: Migration {
    func prepare(on database: Database) -> EventLoopFuture<Void> {
        database.schema(StoredFile.schema)
            .id()
            .field("owner", .string, .required)
            .field("filename", .string, .required)
            .field("path", .string, .required)
            .create()
    }
    func revert(on database: Database) -> EventLoopFuture<Void> {
        database.schema(StoredFile.schema).delete()
    }
}

app.migrations.add(CreateUser())
app.migrations.add(CreateNote())
app.migrations.add(CreateFile())

// run migrations
try app.autoMigrate().wait()

// Insert default admin user plaintext (for testing) if none exists
_ = User.query(on: app.db).count().flatMap { cnt in
    if cnt == 0 {
        return User(username: "admin", password: "123").create(on: app.db)
    } else {
        return app.eventLoopGroup.next().makeSucceededVoidFuture()
    }
}.wait()

// ---------- Safety middleware ----------

// Helper to check loopback address
func isLocalAddress(_ addr: SocketAddress?) -> Bool {
    guard let addr = addr else { return false }
    switch addr {
    case .v4(let v4):
        return v4.address == .loopback
    case .v6(let v6):
        return v6.address == .loopback
    default:
        return false
    }
}

// Middleware: require localhost & token for non-root endpoints
struct LocalOnlyMiddleware: AsyncMiddleware {
    func respond(to request: Request, chainingTo next: AsyncResponder) async throws -> Response {
        // Always allow root (/) so token can be displayed
        if request.url.path == "/" {
            return try await next.respond(to: request)
        }
        // Check remote is loopback
        let remote = request.remoteAddress
        if !isLocalAddress(remote) {
            return Response(status: .forbidden, body: .init(string: "Access denied: local-only"))
        }
        // Require token query param to match env token
        guard let token = request.query[String.self, at: "token"], token == vulnToken else {
            return Response(status: .forbidden, body: .init(string: "Access denied: token missing or invalid. See / for token."))
        }
        return try await next.respond(to: request)
    }
}
app.middleware.use(LocalOnlyMiddleware())

// ---------- Routes (vulnerable endpoints) ----------

// Root: show token and links
app.get { req -> Response in
    let token = vulnToken
    var html = "<h2>VulnLab (Vapor) — Local Only</h2>"
    html += "<p><strong>Auto token:</strong> <code>\(token)</code></p>"
    html += "<p><a href='/home?token=\(token)'>Open Lab Home</a></p>"
    html += "<ul>"
    html += "<li><a href='/notes?token=\(token)'>A01 Broken Access Control (notes)</a></li>"
    html += "<li><a href='/crypto?token=\(token)'>A02 Cryptographic Failures</a></li>"
    html += "<li><a href='/login?user=admin&pass=123&token=\(token)'>A03 SQL Injection (login)</a></li>"
    html += "<li><a href='/transfer?from=a&to=b&amount=9999&token=\(token)'>A04 Insecure Transfer</a></li>"
    html += "<li><a href='/config?token=\(token)'>A05 Misconfiguration / Info Leak</a></li>"
    html += "<li><a href='/upload?token=\(token)'>Unrestricted Upload</a></li>"
    html += "<li><a href='/cmd?c=whoami&token=\(token)'>Command Exec</a></li>"
    html += "<li><a href='/fetch?url=http://127.0.0.1:8080/config&token=\(token)'>SSRF Fetch</a></li>"
    html += "</ul>"
    return Response(status: .ok, body: .init(string: html))
}

// Home
app.get("home") { req -> Response in
    let t = req.query[String.self, at: "token"] ?? vulnToken
    var html = "<h3>Vuln Lab Home</h3><ul>"
    html += "<li><a href='/notes?token=\(t)'>Notes</a></li>"
    html += "<li><a href='/crypto?token=\(t)'>Crypto</a></li>"
    html += "<li><a href='/login?user=admin&pass=123&token=\(t)'>Login (SQLi)</a></li>"
    html += "<li><a href='/upload?token=\(t)'>Upload</a></li>"
    html += "</ul>"
    return Response(status: .ok, body: .init(string: html))
}

// A01 - Broken ACL: returns all notes (no owner filtering) and shows raw content (stored XSS)
app.get("notes") { req async throws -> Response in
    let notes = try await Note.query(on: req.db).all()
    var sb = "<h3>All Notes (no ACL)</h3>"
    for n in notes {
        // intentionally not escaped -> stored XSS possible
        sb += "<div style='border:1px solid #ccc;margin:6px;padding:6px'><strong>Owner:</strong> \(n.owner)<br/><strong>Content:</strong> \(n.content)</div>"
    }
    return Response(status: .ok, body: .init(string: sb))
}

// Add note - vulnerable to stored XSS and SQLi when using raw queries elsewhere
app.post("addnote") { req async throws -> Response in
    // read form fields without validation
    struct F: Content { var owner: String?; var content: String? }
    let f = try req.content.decode(F.self)
    let owner = f.owner ?? "anonymous"
    let content = f.content ?? ""
    // insecure: use raw SQL to simulate concatenation vulnerability
    try await req.db.raw("INSERT INTO notes(owner, content) VALUES ('\(owner)', '\(content)')").run()
    return Response(status: .ok, body: .init(string: "Note added (stored XSS possible)"))
}

// A02 Crypto - plaintext secret + weak AES (predictable IV)
app.get("crypto") { req -> Response in
    let secret = "PlaintextSecret"
    // weak AES-128-CBC with predictable zero IV for demo
    let key = Array("weakkeyweakkey".utf8) // 16 bytes
    let iv = [UInt8](repeating: 0, count: 16)
    var encryptedHex = "encrypt-failed"
    do {
        let ctx = try AES.CBC.encrypt(Array(secret.utf8), key: key, iv: iv)
        encryptedHex = ctx.map { String(format: "%02x", $0) }.joined()
    } catch {
        encryptedHex = "encryption-error"
    }
    let html = "<h4>Plaintext: \(secret)</h4><p>Weak AES-128-CBC (predictable IV) hex: \(encryptedHex)</p>"
    return Response(status: .ok, body: .init(string: html))
}

// A03 SQLi - insecure login (concatenated SQL)
app.get("login") { req async throws -> Response in
    let user = req.query[String.self, at: "user"] ?? ""
    let pass = req.query[String.self, at: "pass"] ?? ""
    // insecure raw SQL concatenation -> SQL injection
    let rows = try await req.db.raw("SELECT COUNT(*) as c FROM users WHERE username = '\(user)' AND password = '\(pass)'").all()
    var count = 0
    if let first = rows.first, let c = first.column("c")?.int {
        count = c
    }
    return Response(status: .ok, body: .init(string: count == 1 ? "Login ok" : "Invalid"))
}

// A04 Insecure design - transfer without auth/limits
app.get("transfer") { req -> Response in
    let from = req.query[String.self, at: "from"] ?? "a"
    let to = req.query[String.self, at: "to"] ?? "b"
    let amount = req.query[String.self, at: "amount"] ?? "0"
    return Response(status: .ok, body: .init(string: "Transferred \(amount) from \(from) to \(to)"))
}

// A05 Info leak - environment + stack trace
app.get("config") { req -> Response in
    var out = "<h4>Environment</h4><pre>"
    for (k,v) in ProcessInfo.processInfo.environment { out += "\(k)=\(v)\n" }
    out += "</pre>"
    out += "<h4>Stack sample</h4><pre>"
    out += Thread.callStackSymbols.prefix(10).joined(separator: "\n")
    out += "</pre>"
    return Response(status: .ok, body: .init(string: out))
}

// A06 Components (simulated)
app.get("components") { req -> Response in
    Response(status: .ok, body: .init(string: "Simulated outdated components (demo)"))
}

// A07 Register - store plaintext password (and set insecure cookie)
app.post("register") { req async throws -> Response in
    struct F: Content { var user: String?; var pass: String? }
    let f = try req.content.decode(F.self)
    guard let user = f.user, let pass = f.pass else {
        return Response(status: .badRequest, body: .init(string: "user/pass required"))
    }
    // insecure raw insert concatenation -> SQLi possible
    try await req.db.raw("INSERT INTO users(username, password) VALUES ('\(user)', '\(pass)')").run()
    // insecure cookie (no HttpOnly/Secure)
    var res = Response(status: .ok, body: .init(string: "User created (plaintext password stored)"))
    res.headers.add(name: .setCookie, value: "lab_user=\(user); Path=/")
    return res
}

// A08 Update (no signature verification)
app.get("update") { req -> Response in
    Response(status: .ok, body: .init(string: "Downloading updates from unverified source..."))
}

// A09 Logging/monitoring failure - return exception info
app.get("log") { req -> Response in
    do {
        throw Abort(.internalServerError, reason: "debug-exception")
    } catch {
        return Response(status: .internalServerError, body: .init(string: "Exception: \(error)"))
    }
}

// A10 SSRF - fetch arbitrary URL with no validation
app.get("fetch") { req -> Response in
    guard let urlStr = req.query[String.self, at: "url"], !urlStr.isEmpty else {
        return Response(status: .badRequest, body: .init(string: "Provide ?url="))
    }
    // insecure fetch with no whitelist
    do {
        let url = URL(string: urlStr)!
        let data = try Data(contentsOf: url)
        if let s = String(data: data, encoding: .utf8) {
            // reflect unsafely (possible XSS)
            return Response(status: .ok, body: .init(string: s))
        } else {
            return Response(status: .ok, body: .init(string: "Fetched binary data"))
        }
    } catch {
        return Response(status: .ok, body: .init(string: "Fetch failed: \(error.localizedDescription)"))
    }
}

// Upload & download endpoints (unrestricted upload, insecure download)
let uploadsDir = app.directory.workingDirectory + "uploads"
try? FileManager.default.createDirectory(atPath: uploadsDir, withIntermediateDirectories: true)

// GET upload form
app.get("upload") { req -> Response in
    let t = req.query[String.self, at: "token"] ?? vulnToken
    let html = """
    <h3>Upload (Unrestricted)</h3>
    <form method="post" enctype="multipart/form-data" action="/upload?token=\(t)">
      Owner: <input name="owner"/><br/>
      File: <input type="file" name="file"/><br/>
      <button>Upload</button>
    </form>
    <p>Uploaded files:</p>
    """
    // list files (no auth)
    var list = "<ul>"
    let files = try? FileManager.default.contentsOfDirectory(atPath: uploadsDir)
    files?.forEach { f in list += "<li>\(f) - <a href='/download?name=\(f)&token=\(t)'>download</a></li>" }
    list += "</ul>"
    return Response(status: .ok, body: .init(string: html + list))
}

// POST upload
app.on(.POST, "upload", body: .collect(maxSize: "50mb")) { req -> EventLoopFuture<Response> in
    struct F: Content { var owner: String? }
    let owner = (try? req.content.get(String.self, at: "owner")) ?? "anonymous"
    // obtain file part unsafely
    guard let part = req.body.data else {
        return req.eventLoop.makeSucceededFuture(Response(status: .badRequest, body: .init(string: "file missing")))
    }
    // naive write - uses no filename from multipart parsing for brevity; this demo writes one static file
    let filename = "upload-\(UUID()).bin"
    let dest = uploadsDir + "/" + filename
    do {
        try part.write(to: URL(fileURLWithPath: dest))
        // store in DB (insecure SQL concat)
        try? req.db.raw("INSERT INTO files(owner, filename, path) VALUES ('\(owner)', '\(filename)', '\(dest)')").run()
    } catch {
        return req.eventLoop.makeSucceededFuture(Response(status: .internalServerError, body: .init(string: "write failed: \(error)")))
    }
    return req.eventLoop.makeSucceededFuture(Response(status: .ok, body: .init(string: "Uploaded: \(filename)")))
}

// Download (no auth / IDOR)
app.get("download") { req -> Response in
    guard let name = req.query[String.self, at: "name"], !name.isEmpty else {
        return Response(status: .badRequest, body: .init(string: "Provide ?name="))
    }
    let path = uploadsDir + "/" + (name as String)
    if FileManager.default.fileExists(atPath: path) {
        let data = try Data(contentsOf: URL(fileURLWithPath: path))
        var res = Response(status: .ok, body: .init(data: data))
        res.headers.replaceOrAdd(name: .contentType, value: "application/octet-stream")
        res.headers.replaceOrAdd(name: .contentDisposition, value: "attachment; filename=\"\(name)\"")
        return res
    } else {
        return Response(status: .notFound, body: .init(string: "file not found"))
    }
}

// Command exec (RCE)
app.get("cmd") { req -> Response in
    guard let c = req.query[String.self, at: "c"], !c.isEmpty else {
        return Response(status: .badRequest, body: .init(string: "Provide ?c="))
    }
    // insecure: executes command via /bin/sh -c
    let process = Process()
    process.executableURL = URL(fileURLWithPath: "/bin/sh")
    process.arguments = ["-c", c]
    let pipeOut = Pipe(); let pipeErr = Pipe()
    process.standardOutput = pipeOut; process.standardError = pipeErr
    do {
        try process.run()
        process.waitUntilExit()
        let out = String(data: pipeOut.fileHandleForReading.readDataToEndOfFile(), encoding: .utf8) ?? ""
        let err = String(data: pipeErr.fileHandleForReading.readDataToEndOfFile(), encoding: .utf8) ?? ""
        return Response(status: .ok, body: .init(string: "OUT:\n\(out)\nERR:\n\(err)"))
    } catch {
        return Response(status: .internalServerError, body: .init(string: "Exec failed: \(error)"))
    }
}

// Reflected XSS (echo)
app.get("echo") { req -> Response in
    let q = req.query[String.self, at: "q"] ?? ""
    // intentionally unsanitized echo
    return Response(status: .ok, body: .init(string: "You said: \(q)"))
}

// Search (SQLi via LIKE)
app.get("search") { req async throws -> Response in
    let q = req.query[String.self, at: "q"] ?? ""
    var sb = "<h3>Search results</h3>"
    // insecure concatenated SQL
    let rows = try await req.db.raw("SELECT username FROM users WHERE username LIKE '%\(q)%'").all()
    for r in rows {
        if let col = r.column("username")?.string {
            sb += "\(col)<br/>"
        }
    }
    return Response(status: .ok, body: .init(string: sb))
}

// Simple add note form (GET)
app.get("addnoteform") { req -> Response in
    let t = req.query[String.self, at: "token"] ?? vulnToken
    let html = """
    <form method="post" action="/addnote?token=\(t)">
      Owner: <input name="owner"/><br/>
      Content: <textarea name="content"></textarea><br/>
      <button type="submit">Add Note</button>
    </form>
    """
    return Response(status: .ok, body: .init(string: html))
}

// Start app
try app.run()

// ---------- Minimal AES helper (CBC) ----------
enum AES {
    enum CBC {
        static func encrypt(_ data: [UInt8], key: [UInt8], iv: [UInt8]) throws -> [UInt8] {
            // use CommonCrypto via CryptoKit is complicated here; do a naive AES using Crypto if available
            // For demo, we'll try to use CryptoKit AES.GCM as fallback to produce some bytes, but keep semantics simple
            // NOTE: This is only an educational placeholder; not used for security.
            if #available(macOS 10.15, *) {
                import CryptoKit
            }
            // fallback: XOR with key (weak "encryption") to illustrate predictable weak cipher
            var out = [UInt8](repeating: 0, count: data.count)
            for i in 0..<data.count {
                out[i] = data[i] ^ key[i % key.count]
            }
            return out
        }
    }
}
